composer_migration/src/composer_migration/lib/comparator.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2023 Google LLC
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     https://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, TypeVar
+import logging
+import os
+
+from composer_migration.lib.strategies.strategy import DAGChecker
+
+from rich.console import Console
+from rich.table import Table
+from rich.console import Group
+from rich.text import Text
+from rich.panel import Panel
+
+# https://peps.python.org/pep-0484/#annotating-instance-and-class-methods
+T = TypeVar("T", bound="DAGComparator")
+
+
+class DAGComparator:
+    def __init__(self: T, dags_directory: str) -> None:
+        self.problem_operators: List[str] = []
+        self.dags_directory = dags_directory
+
+    def check_dag_files(self: T, strategy: DAGChecker) -> None:
+        dags_dir = os.listdir(self.dags_directory)
+        console = Console()
```

Comment:
suggestion. split out presentation logic from dag file checking logic. Down the line you may want to output a JSON file with errors found, instead of just a table. You could keep check_dag_files(self: T, strategy: DAGchecker) -> None & keep storing issues found in self.problem_operators.   As a suggestion you could refactor all the rich table code into another function present_issues_to_cli(self: T) -> None which has all the logic to print to console & uses the state in self.problems operators. Down the line you may want to write write_issues_to_json(self: t) etc. This is just a suggestion !

Reply:
this is a good suggestion. IIRC, I think I originally wanted to go down this path, then got tripped up with something syntax related and abandoned it in favor of getting it working. It's time to pick this up again because of the reasons you've highlighted

Reply:
Update: I do think I want separate tables for each check, but I think that the logic should still be split